components/file_system/src/rate_limiter.rs

```diff
@@ -1,133 +1,362 @@
+        request_impl!(self, priority, amount, "sync")
+    }
+
+    async fn async_request(&self, priority: IOPriority, amount: usize) -> usize {
+        request_impl!(self, priority, amount, "async")
+    }
+
+    // Called by a daemon thread every `DEFAULT_REFILL_PERIOD`.
+    // It is done so because the algorithm correctness relies on refill epoch being faithful to physical time.
+    fn refill(&self) {
+        let locked = self.protected.try_write_for(DEFAULT_REFILL_PERIOD);
+        if locked.is_none() {
+            panic!("Can't acquire lock to refill IO rate limiter");
+        }
+        let mut locked = locked.unwrap();
+
+        let mut limit = self.ios_per_sec[IOPriority::High as usize].load(Ordering::Relaxed);
+        if limit == 0 {
+            return;
         }
-        for i in self.write.iter() {
-            i.store(0, Ordering::Relaxed);
+
+        let now = Instant::now_coarse();
+        locked.last_refill_time = now;
+
+        let mut ios_through = self.ios_through[IOPriority::High as usize].swap(
+            self.pending_ios[IOPriority::High as usize].swap(0, Ordering::Relaxed),
+            Ordering::Release,
+        );
+        for p in &[IOPriority::Medium, IOPriority::Low] {
+            let pi = *p as usize;
+            limit = if limit > ios_through {
+                limit - ios_through
+            } else {
+                1 // a small positive value
+            };
+            self.ios_per_sec[pi].store(limit, Ordering::Relaxed);
+            ios_through = self.ios_through[pi].swap(
+                self.pending_ios[pi].swap(0, Ordering::Relaxed),
+                Ordering::Release,
+            );
         }
     }
 }
 
-/// No-op limiter
+#[derive(Debug)]
+struct Refiller {
+    sender: Option<Mutex<Sender<()>>>,
+    handle: Option<JoinHandle<()>>,
+}
+
+impl Refiller {
+    pub fn new() -> Self {
+        Refiller {
+            sender: None,
+            handle: None,
+        }
+    }
+
+    pub fn must_start(&mut self, limiter: Arc<RawIORateLimiter>) {
+        let (tx, rx) = mpsc::channel();
+        let h = ThreadBuilder::new()
+            .name("refiller".to_owned())
+            .spawn(move || {
+                tikv_alloc::add_thread_memory_accessor();
+                while let Err(mpsc::RecvTimeoutError::Timeout) =
+                    rx.recv_timeout(DEFAULT_REFILL_PERIOD)
+                {
+                    limiter.refill();
+                }
+                tikv_alloc::remove_thread_memory_accessor();
+            })
+            .unwrap();
+        self.handle = Some(h);
+        self.sender = Some(Mutex::new(tx));
+    }
+}
+
 /// An instance of `IORateLimiter` should be safely shared between threads.
 #[derive(Debug)]
 pub struct IORateLimiter {
-    refill_bytes: usize,
-    enable_statistics: bool,
+    priority_map: [IOPriority; IOType::VARIANT_COUNT],
+    bytes: Arc<RawIORateLimiter>,
+    // TODO: 1) move this to background pool. 2) pause it when limit is toggled off.
```

Comment:
also 3. stop on server shutdown?

Reply:
I think it will be shutdown when Limiter is dropped, or managed by background pool.